src/util/db/dbid.h

```diff
@@ -1,126 +1,138 @@
-#include <ostream>
-#include <utility>
-
 #include <QString>
 #include <QVariant>
 #include <QtDebug>
+#include <functional>
+#include <ostream>
+#include <utility>
 
 #include "util/assert.h"
 
-
-// Base class for ID values of objects that are stored in the database.
+/// Base class for ID values of objects that are stored in the database.
 //
-// Implicit conversion from/to the native value type has been disabled
-// on purpose in favor of explicit conversion. The internal value type
-// is also hidden as an implementation detail.
+/// Implicit conversion from/to the native value type has been disabled
+/// on purpose in favor of explicit conversion. The internal value type
+/// is also hidden as an implementation detail.
 //
-// Although used as a base class this class has a non-virtual destructor,
-// because destruction is trivial and for maximum efficiency! Derived
-// classes may add additional behavior (= member functions), but should
-// not add any additional state (= member variables). Inheritance is
-// only needed for type-safety.
+/// Although used as a base class this class has a non-virtual destructor,
+/// because destruction is trivial and for maximum efficiency! Derived
+/// classes may add additional behavior (= member functions), but should
+/// not add any additional state (= member variables). Inheritance is
+/// only needed for type-safety.
 class DbId {
-protected:
-public:
-    // Alias for the corresponding native type. It keeps the
-    // implementation of this class flexible if we ever gonna
-    // need to change it from 'int' to 'long' or any other type.
+  public:
+    /// Alias for the corresponding native type. It keeps the
+    /// implementation of this class flexible if we ever gonna
+    /// need to change it from 'int' to 'long' or any other type.
     typedef int value_type;
 
-    DbId()
-        : m_value(kInvalidValue) {
-        DEBUG_ASSERT(!isValid());
+    static constexpr value_type kInvalidValue = -1;
+    static constexpr value_type kMinValue = 0;
+
+    static constexpr bool isValidValue(value_type value) {
+        static_assert(kInvalidValue < kMinValue);
+        return value >= kMinValue;
+    }
+
+    /// Type tag for overloaded constructor that doesn't assert
+    struct NoAssert {};
+
+    constexpr DbId()
+            : m_value(kInvalidValue) {
+    }
+    constexpr DbId(NoAssert, value_type value)
+            : m_value(value) {
     }
     explicit DbId(value_type value)
-        : m_value(value) {
-        DEBUG_ASSERT(isValid() || (kInvalidValue == m_value));
+            : DbId(NoAssert{}, value) {
+        DEBUG_ASSERT(isValid() || m_value == kInvalidValue);
     }
     explicit DbId(QVariant variant)
-        : DbId(valueOf(std::move(variant))) {
+            : DbId(NoAssert{}, valueOf(std::move(variant))) {
     }
 
-    bool isValid() const {
+    constexpr bool isValid() const {
         return isValidValue(m_value);
     }
 
+    constexpr value_type valueNoAssert() const {
```

Comment:
valueMaybeInvalid() 

We use this pattern in the FramePos class and it explains the usage well, even if assertion are disabled.